operator/pkg/ciliumendpointbatch/endpointbatch.go

```diff
@@ -0,0 +1,272 @@
+	aggregator *aggregator
+
+	// reconciler is an util used to reconcile CiliumEndpointBatch changes.
+	reconciler *reconciler
+
+	// Manager is used to create and maintain a local datastore. Manager watches for
+	// cilium endpoint changes and enqueues/dequeues the cilium endpoint changes in CEB.
+	// It maintains the desired state of the CEBs in dataStore
+	Manager cebManager
+
+	// cebDeleteSyncTime is used track latency between CEBatchDeleteSyncPeriod. if the latency
+	// exceeds CEBatchDeleteSyncPeriod it will reconcile all CEBs [create, update and delete]
+	// with the k8s-apiServer.
+	cebDeleteSyncTime time.Duration
+
+	// ciliumEndpointStore is used to get a list of current active CEPs in a cluster. By, using this cache
+	// we can avoid costly api calls to the k8s-apiserver.
+	ciliumEndpointStore cache.Indexer
+
+	// ciliumEndpointBatchUpsertSyncPeriod indicates the minimum delay period for
+	// some CiliumEndpointBatches to sync with the k8s-apiserver. Newly created and
+	// updated CiliumEndpointBatches are synced after this interval.
+	// The cases for syncing Update CiliumEndpointBatches are
+	// 1) Insert new CEPs in a CEB.
+	// 2) Modify existing CEPs in a CEB.
+	ciliumEndpointBatchUpsertSyncPeriod time.Duration
+
+	// ciliumEndpointBatchDeleteSync indicates the minimum delay period for syncing
+	// some CiliumEndpointBatches with the k8s-apiserver. Deleted CEB's and
+	// updated CiliumEndpointBatches are synced after this interval.
+	// The case for syncing Update CiliumEndpointBatches is
+	// 1) Remove CEPs in a CEB.
+	ciliumEndpointBatchDeleteSyncPeriod time.Duration
+}
+
+var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "ciliumEndpointBatch")
+
+// Derives the unique name from CoreCiliumEndpoint object.
+// This unique name is used for mapping CiliumEndpoint to CiliumEndpointBatch.
+// Used widely, to determine if the given CEP is mapped to any CEB or not.
+func getCepNameFromCCEP(cep *capi_v2a1.CoreCiliumEndpoint) string {
+	return cep.Namespace + "/" + cep.Name
+}
+
+// NewCebController, creates and initializes the CEB controller
+func NewCebController(client *k8s.K8sCiliumClient,
+	cebSyncPeriod time.Duration,
+	cebDeleteSyncPeriod time.Duration,
+	maxCepsInCeb int,
+) *CiliumEndpointBatchController {
+
+	aggregator := newAggregator()
+
+	manager := newCebManagerFcfs(aggregator, maxCepsInCeb)
+
+	cebC := &CiliumEndpointBatchController{
+		clientV2:                            client.CiliumV2(),
+		clientV2a1:                          client.CiliumV2alpha1(),
+		aggregator:                          aggregator,
+		reconciler:                          newReconciler(client.CiliumV2alpha1(), manager),
+		Manager:                             manager,
+		cebDeleteSyncTime:                   0,
+		ciliumEndpointBatchUpsertSyncPeriod: cebSyncPeriod,
+		ciliumEndpointBatchDeleteSyncPeriod: cebDeleteSyncPeriod,
+	}
+
+	// List all existing CEBs from the API server and cache it locally.
```

Comment:
You're absolutely spot on constructors. Here is the current workflow for the CEB controller,
1) Create the CEB controller
2) Sync the current CEBs from api-sever to local cache.
3) Initialize and run CEP watcher
4) Run Ceb controller.
On operator restart/warm boot. We need to sync the current CEBs from api-server  to local-cache before we start the watcher. 
Instead if we start and run CEP watcher first, all existing CEPs in a cluster would be considered as new `CEP` by CEB controller, because it doesn't any CEB's in local cache yet.

If CEB controller have its own `CEP watcher` that would have made code more cleaner.

Reply:
Ah, thanks for the explanation.
It sounds like we need to wait for the CEP watcher not just to start, but to actually catch up with all existing CEPs, right?

I see you added some more comments below, thanks! Can you also add the fact that the Run() should only be called after the CEP watcher has obtained its initial list of CEPs.